ported to zsh

skywind3000 opened the pull request:
1. rename commacd.bash -> commacd.sh
2. upate README and Makefile
3. support zsh

I have tried my best to minimize modification, here is a list of changed functions:

- _commacd_expand
- _commacd_choose_match
- _commacd_backward_by_prefix

Others are nearly unchanged, except rename local variable "$path" to "$pathname".
because on zsh, assigning a lower case "$path" will override upper case "$PATH", when no_case_glob is enabled.

skywind3000 commented:
I have no idea why there is some error on travis ?

```
commacd.sh:19:1: error: Did you forget the 'then' for this 'if'? [SC1049]
commacd.sh:19:1: error: Couldn't parse this if expression. [SC1073]
commacd.sh:19:30: warning: Use semicolon or linefeed before 'then' (or quote to make it literal). [SC1010]
commacd.sh:21:1: error: Expected 'then'. [SC1050]
commacd.sh:21:6: error: Unexpected keyword/token. Fix any mentioned problems and try again. [SC1072]
```

my code is 

```bash
if [[ -n "$BASH_VERSION" ]]; then
  shopt -s nocaseglob
elif [[ -n "$ZSH_VERSION" ]]; then
  setopt no_case_glob
else
  echo "Sorry, commacd.sh is only available for bash and zsh" >&2
  return
fi
```

It passes all the shellcheck rules on my local computer. but failed on travis.

skywind3000 commented:
`commacd.sh` has also passed all the test cases from shpec on zsh (manually).

shyiko commented:
Thank you, @skywind3000! This is awesome! I'll review & merge over the weekend (don't bother with the failed build, I'll take care of it). Thank you once again!

skywind3000 commented:
Thanks for your review. Because original `,,` command cannot work correctly on WSL/cygwin/msys. I will explain it in detail:

Suppose that there is a directory on Windows:

    D:\Dev\Python36\Lib

And both:

    /mnt/D/Dev/Python36/Lib
    /mnt/d/dev/python36/lib

are valid path names for "D:\Dev\Python36\Lib" on WSL. When I was in the directory of "/mnt/d/dev/python36/lib" , I noticed that:

    ,, pyth

couldn't change the current working directory from "/mnt/d/dev/python36/lib" to "/mnt/d/dev/python3" on both bash and zsh.

### reproduce

step 1. source original `commacd.bash` and cd to "/mnt/d/Dev/Python36/Lib":

![comma-2](https://user-images.githubusercontent.com/3035071/49440609-09ec4380-f7ff-11e8-86d8-2ad787749ee5.png)

As you see, command `,, pyth` works as expected in "/mnt/d/Dev/Python/Lib".

step 2. cd to "/mnt/d/dev/python36/lib" and execute same command again:

![comma-3](https://user-images.githubusercontent.com/3035071/49440764-7b2bf680-f7ff-11e8-91a7-d42c190973eb.png)

it fails and can't switch to its parent directory.

### Investigation

So, I added some log in `_commacd_backward_by_prefix` and located the case sensitive problem:

```bash
_commacd_backward_by_prefix() {
  local prev_dir dir="${PWD%/*}" matches match IFS=$'\n'
  while [[ -n "$dir" ]]; do
    prev_dir="$dir"
    dir="${dir%/*}"
    matches=($(_commacd_expand "$dir/${1}*/"))
    for match in "${matches[@]}"; do
        if [[ "$match" == "$prev_dir/" ]]; then
          echo -n "$prev_dir"
          return
        fi
    done
done
```

According to the code, when running `_commacd_backward_by_prefix pyth` in "/mnt/d/dev/python36/lib", the variable `$dir` is initialed to `${PWD%/*}` which is the parent directory of `$PWD`:

![comma-5](https://user-images.githubusercontent.com/3035071/49441282-03f76200-f801-11e8-9124-f7b9400d0448.png)

In the first iteration of the "while", variable "$prev_dir" and "$dir" will be updated as:

```bash
prev_dir="/mnt/d/dev/python36"
dir="/mnt/d/dev"
```

See the demo:

![comma-6](https://user-images.githubusercontent.com/3035071/49441688-41a8ba80-f802-11e8-9ef5-6218982cd811.png)

Until now, everything works fine, but the result from:

    matches=($(_commacd_expand "$dir/${1}*/"))

is not what I want:

![comma-7](https://user-images.githubusercontent.com/3035071/49442045-3e61fe80-f803-11e8-9d79-674a559ee19f.png)

You see, it returns an upper case "Python36" not the expected "python36", but current `$prev_dir` is 

    /mnt/d/dev/python36

which cannot be matched in `$matches` when using default case sensitive comparison.

As a consequence, `_commacd_backward_by_prefix` failed on WSL. And it will also fail on cygwin/msys.

So I added a `imatch` to handle this and made `commacd` work on WSL/Cygwin/msys.

skywind3000 commented:
where should I add "unsetopt nomatch" ? in the global scope of `commacd.sh` or local to `_commacd_expand` ??

shyiko commented:
Sorry for the delay @skywind3000. I changed the job recently, still "drinking from a firehouse". Nothing is needed from your side. I'm going to merge it as soon as I can.

Commit:
new: enable customizing vcs root markers by $COMMACD_ROOTMARKER

Commit:
Merge branch 'master' of https://github.com/skywind3000/commacd

shyiko commented:
Cherry picked into [1.0.0](https://github.com/shyiko/commacd/blob/master/CHANGELOG.md#100---2019-02-03) 🎉 Thank you, @skywind3000!

skywind3000 commented:
Cool, thank you.
